File: faq/app/data_loader.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from .config import Settings, settings
from .media import ImageRef, extract_image_refs
from .text_utils import answer_summary, stable_hash, strip_markdown


@dataclass(frozen=True)
class FaqItem:
    id: str
    question: str
    variants: list[str]
    answer: str
    enabled: bool
    updated_at: str
    answer_text: str
    answer_summary: str
    images: list[ImageRef]


def _as_str(value: Any, field: str, item_id: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise ValueError(f"{item_id}: `{field}` must be a non-empty string")
    return value.strip()


def _parse_item(raw: dict[str, Any], cfg: Settings) -> FaqItem:
    item_id = _as_str(raw.get("id"), "id", "<unknown>")
    variants_raw = raw.get("variants")
    if not isinstance(variants_raw, list):
        raise ValueError(f"{item_id}: `variants` must be a list")
    variants = [str(v).strip() for v in variants_raw if str(v).strip()]
    answer = _as_str(raw.get("answer"), "answer", item_id)
    return FaqItem(
        id=item_id,
        question=_as_str(raw.get("question"), "question", item_id),
        variants=variants,
        answer=answer,
        enabled=bool(raw.get("enabled", False)),
        updated_at=_as_str(raw.get("updated_at"), "updated_at", item_id),
        answer_text=strip_markdown(answer),
        answer_summary=answer_summary(answer),
        images=extract_image_refs(answer, cfg),
    )


def load_faq_items(cfg: Settings = settings, *, include_disabled: bool = False) -> list[FaqItem]:
    data = json.loads(Path(cfg.data_path).read_text(encoding="utf-8"))
    items_raw = data.get("items")
    if not isinstance(items_raw, list):
        raise ValueError("questions.json must contain an `items` list")
    items = [_parse_item(raw, cfg) for raw in items_raw if isinstance(raw, dict)]
    if not include_disabled:
        items = [item for item in items if item.enabled]
    return items


def data_hash(cfg: Settings = settings) -> str:
    return stable_hash(Path(cfg.data_path).read_text(encoding="utf-8"))

File: Router/server/src/services/rag/dataLoader.js
import fs from "node:fs";
import { extractImageRefs } from "./media.js";
import { answerSummary, stableHash, stripMarkdown } from "./textUtils.js";

function asStr(value, field, itemId) {
    if (typeof value !== "string" || !value.trim())
        throw new Error(`${itemId}: \`${field}\` must be a non-empty string`);
    return value.trim();
}

export function parseItem(raw, assetsDir) {
    const itemId = asStr(raw?.id, "id", "<unknown>");
    const variantsRaw = raw?.variants;
    if (!Array.isArray(variantsRaw))
        throw new Error(`${itemId}: \`variants\` must be a list`);
    const variants = variantsRaw.map((v) => String(v).trim()).filter(Boolean);
    const answer = asStr(raw?.answer, "answer", itemId);
    return {
        id: itemId,
        question: asStr(raw?.question, "question", itemId),
        variants,
        answer,
        enabled: raw?.enabled !== false,
        updated_at: asStr(raw?.updated_at, "updated_at", itemId),
        answer_text: stripMarkdown(answer),
        answer_summary: answerSummary(answer),
        images: extractImageRefs(answer, assetsDir),
    };
}

export function loadFaqItems(filePath, assetsDir, { includeDisabled = false } = {}) {
    const data = JSON.parse(fs.readFileSync(filePath, "utf-8"));
    const itemsRaw = data.items;
    if (!Array.isArray(itemsRaw))
        throw new Error("questions.json must contain an `items` list");
    let items = itemsRaw
        .filter((raw) => raw && typeof raw === "object")
        .map((raw) => parseItem(raw, assetsDir));
    if (!includeDisabled)
        items = items.filter((item) => item.enabled);
    return items;
}

export function dataHashFromContent(content) {
    return stableHash(content);
}

export function dataHashFromFile(filePath) {
    return stableHash(fs.readFileSync(filePath, "utf-8"));
}

export function itemToResult(item, scores = {}) {
    return {
        id: item.id,
        question: item.question,
        answer: item.answer,
        answer_summary: item.answer_summary,
        updated_at: item.updated_at,
        images: item.images,
        variants: item.variants,
        ...scores,
    };
}

export function buildItemMap(items) {
    const map = new Map();
    for (const item of items)
        map.set(item.id, item);
    return map;
}
